packages/client/hmi-client/src/services/model-templating.ts

```diff
@@ -79,16 +77,223 @@ export function junctionCleanUp(modelTemplates: ModelTemplates) {
+					addTemplateArguments.controller_name = controller.target;
+					addTemplateArguments.controller_initial_value = controller.expression;
+
+					// Remove controller from initials
+					initials.splice(index, 1);
+				}
+			}
+
+			// Now that there are no controllers in initals we can add subject/outcome to the arguments
+			if (
+				name === DecomposedModelTemplateTypes.NaturalConversion ||
+				name === DecomposedModelTemplateTypes.ControlledConversion
+			) {
+				// If it's a conversion template, the first two initials are the subject then outcome
+				addTemplateArguments.subject_name = initials[0].target;
+				addTemplateArguments.subject_initial_value = initials[0].expression;
+				addTemplateArguments.outcome_name = initials[1].target;
+				addTemplateArguments.outcome_initial_value = initials[1].expression;
+			} else if (
+				name === DecomposedModelTemplateTypes.NaturalProduction ||
+				name === DecomposedModelTemplateTypes.ControlledProduction
+			) {
+				// If it's a production template, the first initial is the outcome
+				addTemplateArguments.outcome_name = initials[0].target;
+				addTemplateArguments.outcome_initial_value = initials[0].expression;
+			} else {
+				// If it's a degradation template, the first initial is the subject
+				addTemplateArguments.subject_name = initials[0].target;
+				addTemplateArguments.subject_initial_value = initials[0].expression;
+			}
+		} else {
+			const { observables } = modelTemplate.semantics.ode;
+			addTemplateArguments.new_id = observables[0].id;
+			addTemplateArguments.new_name = modelTemplate.header.name;
+			addTemplateArguments.new_expression = observables[0].expression_mathml;
+		}
+
+		kernelManager
+			.sendMessage(`add_${snakeCase(name)}_template_request`, addTemplateArguments)
+			.register(`add_${snakeCase(name)}_template_response`, (d) => {
+				outputCode(d);
+			});
+	}
+	// FIXME: There is some lag when placing a card if these are put in the register callback which feels off
+	// Perhaps there can be some sort of transition state to show the card is being placed?
+	modelTemplate.metadata.templateCard.id = uuidv4();
+	modelTemplates.models.push(modelTemplate);
+}
+
+export function updateDecomposedCardName(
+	modelTemplates: ModelTemplates,
+	kernelManager: KernelSessionManager,
+	outputCode: Function,
+	name: string,
+	id: string
+) {
+	const index = findCardIndexById(modelTemplates, id);
+	const model = modelTemplates.models[index];
+
+	kernelManager
+		.sendMessage('replace_template_name_request', {
+			old_name: model.header.name,
+			new_name: name
+		})
+		.register('replace_template_name_response', (d) => {
+			outputCode(d);
+
+			model.header.name = name;
+			model.metadata.templateCard.name = name;
+			if (model.model.transitions[0] && model.semantics.ode.rates[0]) {
+				model.model.transitions[0].id = name;
+				model.model.transitions[0].properties.name = name;
+				model.semantics.ode.rates[0].target = name;
+			}
+		});
```

Comment:
The way you are using and passing around the kernel manager, now that there are most instance of its usage, is a bit odd. It feels like we should almost invert the pattern because the kernel manager is the wrapper. 

e.g. the pattern I am seeing is:
```
function blah (a, b, c, kernelManager) {
  kernelManagerr.sendMessage(...).on('some_response', () => {
    // do stuff with a, b, c
  });
```
If may be easier to separate responsibilities, and easier to test, if we do

```
function handleResult(a, b, c) { .... }

kernelManager
  .sendMessage(...)
  .on('some_response', () => {
    handleResult( ... )
  });
```

Food for thought, let me know what you think and we should have a chat about it before/if you actually start to change things.

Reply:
Will handle in subsequent PRs